### PR TITLE
POST /api/v1/users/login

### PR DESCRIPTION
**Pull request by Alaalser**
Relates #9
## POST /api/v1/users/login
- Created a query to get user data to check if the user existed and to compare the passwords
- Made a validation on email and password.
- Built the login controller which does the validation, compares passwords, and handles any errors.
- Run tests on the query and the route.

### CODE REVIEW

File: server/controllers/users/login/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    const {
+      rows: [user],
+    } = await getUserDataQeury({ email });
+
+    if (user) {
+      const { id: userId, password: hashedPassword } = user;
+      const comparedPasswords = await compare(password, hashedPassword);
+      if (!comparedPasswords) {
+        return next(boomify(400, 'Login Error', 'Incorrect password'));
```
Comment (AlaaShurrab) [inline server/controllers/users/login/index.js:20]: change the error to say incorrect email or password

File: server/utils/validation/loginSchema/index.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const { string, object } = require('yup');
+
+const loginSchema = object({
+  email: string().email('Must be a valid email').required('Email is required'),
+  password: string()
+    .min(8, 'Password must be at least 8 char')
+    .required('Password is required'),
```
Comment (AlaaShurrab) [inline server/utils/validation/loginSchema/index.js:7]: the yup error handler provaied the exact same massage so why are you writing it this way?

File: server/controllers/users/login/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      }
+      const token = await sign({ userId });
+      return res
+        .cookie('token', token, { httpOnly: true })
+        .status(201)
+        .json({ statusCode: 201, message: 'logged in successfully' });
+    }
+    return next(boomify(400, 'Login Error', 'Incorrect email'));
```
Comment (AlaaShurrab) [inline server/controllers/users/login/index.js:28]: change the error to say incorrect email or password

File: server/controllers/users/login/index.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    if (user) {
+      const { id: userId, password: hashedPassword } = user;
+      const comparedPasswords = await compare(password, hashedPassword);
+      if (!comparedPasswords) {
+        return next(boomify(400, 'Login Error', 'Incorrect password'));
+      }
+      const token = await sign({ userId });
+      return res
+        .cookie('token', token, { httpOnly: true })
+        .status(201)
```
Comment (hassanelnajjar) [inline server/controllers/users/login/index.js:25]: Status code should be 200

File: server/controllers/users/login/index.js (hunk 3 above)
Comment (hassanelnajjar) [inline server/controllers/users/login/index.js:28]: Please check error first

File: server/controllers/users/login/index.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (user) {
+      const { id: userId, password: hashedPassword } = user;
+      const comparedPasswords = await compare(password, hashedPassword);
+      if (!comparedPasswords) {
+        return next(boomify(400, 'Login Error', 'Incorrect password'));
+      }
+      const token = await sign({ userId });
+      return res
+        .cookie('token', token, { httpOnly: true })
```
Comment (hassanelnajjar) [inline server/controllers/users/login/index.js:24]: add secure: process.env.NODE_ENV === 'production' to cookies

**Review by hassanelnajjar — DISMISSED** [review]

**Review by hassanelnajjar — APPROVED** [review]

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by AlaaShurrab — APPROVED** [review]

**Pull request merged**